### ISSUE TITLE
Wildcard Routes not matching correctly

### ISSUE THREAD

**Issue by AStoker** [aurelia/route-recognizer#22]
### Help Wanted/Possible Enhancement  
Initially posted here: https://github.com/aurelia/router/issues/351
The problem seems to be related more specifically to the route-recognizer.
Repro Repo: https://github.com/AStoker/aurelia-route-recognizer-issue

I have an application that consists of two "main" pages, "Home" and "Settings". For both of those pages, there are child routes.  Both Home and Settings define their main route in the `app.js`, and then the subsequent child routes are defined in each view model (`home.js` and `settings.js`).   

Home does not have any prepended route to navigate to sub routes. So if the user wanted to go to the inbox, the route would simply be `/inbox`. But for the Settings page, all routes are prepended with `/settings`. So if the user wanted to navigate to a subpage in Settings, it would have the route `/settings/flow`.  

My routes are as follows:
``` javascript
//App Route Config
       {
            route: ['settings', 'settings*details'],
            name: 'Settings',
            title: 'Settings',
            moduleId: 'settings/settings',
            href: '#settings',
            nav: true
        },
        {
            route: ['','*details'],
            name: 'Home',
            title: 'Home',
            moduleId: 'central/central',
            href: '#',
            nav: true
        }
```
``` javascript
//Settings Route Config
       {
            route: [''],
            name: 'SettingsHome',
            title: 'Home',
            moduleId: 'empty/empty',
            nav: false,
        },
        {
            route: ['flow'],
            name: 'FlowSettings',
            title: 'Workflow',
            href: '#/settings/flow',
            moduleId: 'settings/flow/flowList',
            nav: true,
        },
        {
            route: ['forms'],
            name: 'FormSettings',
            title: 'Forms',
            href: '#/settings/forms',
            moduleId: 'settings/form/formList',
            nav: true,
        },
```
``` javascript
// Home Route Config
      {
            route: [''],
            name: 'Central',
            title: 'Home',
            viewPorts: {
                'central-nav': 	{
                    moduleId: 'central/navigation',
                },
                'sub-nav': 		{
                    moduleId: 'empty/empty',
                },
                'work-space': 	{
                    moduleId: 'workspace/workspace',
                }
            },
            nav: false,
        },
        {
            route: ['inbox', 'inbox/:workQueueId'],
            name: 'Inbox',
            title: 'Inbox',
            viewPorts: {
                'central-nav': 	{
                    moduleId: 'central/navigation'
                },
                'sub-nav': 		{
                    moduleId: 'flow/inbox/inbox'
                },
                'work-space': 	{
                    moduleId: 'workspace/workspace'
                }
            },
            nav: true,
        },
```
Now, with all those code examples out of the way, the trouble with these routes are that when I try to navigate to a sub page in Settings, the route matching does not match correctly to the Settings route/subroute, but tries to fall back to the Home route.  
So when I navigate to `#/settings/flow`, the router matches it to Home route via `*details` instead of `settings*details`.  
I tried going through the router matching logic, and I'm slowly picking up how it's doing its stuff. I had to put the Settings route first in the list to make it match `settings*details` before `*details` (at least it seemed that I had to). But as it goes down into the matching, it fails to start matching during the `*details` stage of the Settings route.  

So I guess it comes down to this. I'm either doing something wrong, and would greatly appreciate the help, or the routing is broken in this particular use case and an enhancement should be put in place to help address it.  Any thoughts?

### TL;DR
I'm trying to get my routes hooked up so the root can route to Child-A via `*details`, and Child-B via `child*details`. This works fine for all the routes and child routes in Child-A, but Child-B can't get it's child routes working since they all try to match to the root route config of `*details`.

**Comment by AStoker** [comment]
@EisenbergEffect, I'd really appreciate some direction on this one. I've been trying to dig into the recognizer code to understand how all the pieces work together, and then maybe make a PR to fix the issue. But I'm having a hard time pinning down where I should be looking for solving this issue. Any ideas?

**Comment by EisenbergEffect** [comment]
@bryanrsmith is the best person to guide you on this.

**Comment by AStoker** [comment]
Thanks, I appreciate it.

**Comment by bryanrsmith** [comment]
Matching happens on the entire path segment. You can't use wildcards to match segment prefixes. It's probably trying to match the static string `"settings*details"` (like the URL would need to be `#/settings*details`). Try `settings/*details`, which would match the "settings" static segment, followed by a wildcard segment named "details".

**Comment by AStoker** [comment]
Well, that seems to work, so first let me just say thank you for saving me from more hours of head to keyboard banging.  
However, how is this different from wildcard routes as defined by the docs (http://aurelia.io/docs.html#/aurelia/framework/1.0.0-beta.1.2.4/doc/article/cheat-sheet/7)?

**Comment by bryanrsmith** [comment]
Hmm. Those docs seem wrong to me. I can't recall it ever working with that
syntax. Perhaps I'm misremembering or it used to work like that before I
joined? We should probably update the docs to the current syntax.

**Comment by AStoker** [comment]
Okay, I just wanted to make sure I understood what was going on. Thanks for the clarification. And yah, sounds like the docs need to be updated. Thanks again!
